src/asm_files.cpp

```diff
@@ -36,15 +36,34 @@ int create_map_crab(const EbpfMapType& map_type, uint32_t key_size, uint32_t val
     return cache.at(equiv);
 }
 
-EbpfMapDescriptor* find_map_descriptor(int map_fd) {
-    for (EbpfMapDescriptor& map : global_program_info.map_descriptors) {
+static EbpfMapDescriptor* find_map_descriptor(program_info& info, int map_fd) {
+    for (EbpfMapDescriptor& map : info.map_descriptors) {
         if (map.original_fd == map_fd) {
             return &map;
         }
     }
     return nullptr;
 }
 
+EbpfMapDescriptor* find_map_descriptor(int map_fd) {
+    return find_map_descriptor(global_program_info, map_fd);
+}
+
+static void validate_map_in_map(const ebpf_platform_t* platform, program_info& info) {
+    int i = 0;
+    for (const EbpfMapDescriptor& outer_map : info.map_descriptors) {
+        if (platform->get_map_type(outer_map.type).value_type == EbpfMapValueType::MAP) {
+            const EbpfMapDescriptor* p_inner_map = find_map_descriptor(info, outer_map.inner_map_fd);
+            assert(p_inner_map);
+            if (p_inner_map->key_size != outer_map.value_size) {
+                throw std::runtime_error(std::string("value_size of map number ") + std::to_string(i) +
+                                         " and key_size for its inner map are different.");
```

Comment:
My understanding from https://hechao.li/2019/03/19/Use-Map-in-Map-in-BPF-programs-via-Libbpf/ was that the key size of the inner map is irrelevant, the value size of the outer map must always be sizeof(uint32_t) just like for BPF_MAP_TYPE_PROG_ARRAY.   The value for prog array is a prog id, and the value for HASH_OF_MAPS and ARRAY_OF_MAPS is a map id.   And those ids are usable with libbpf apis like:
```
LIBBPF_API int bpf_prog_get_fd_by_id(__u32 id);
LIBBPF_API int bpf_map_get_fd_by_id(__u32 id);
```
So the application can turn them into fds for the prog and map, respectively.
Let me know if you find otherwise.

Reply:
You are right. My bad - there is absolutely no reason for those to match.